src/storage/BlockStorage.ts

```diff
@@ -0,0 +1,295 @@
+
+    /*────────────────────────────────────────────────────────────────────────────
+      IMPLEMENTATION
+    ────────────────────────────────────────────────────────────────────────────*/
+    deleteBlock(hashOrForkId: Hash | ForkId, height?: BlockHeight): boolean {
+        if (height === undefined) {
+            // ┌─ ROUTES TO: [OVERLOAD 1] - delete by hash
+            const blockConfirmation = this.hashToBlockMap.get(
+                hashOrForkId as Hash
+            );
+            if (!blockConfirmation) return false;
+
+            // Need to find and delete from coordinates map too
+            const block = Block.decode(
+                blockConfirmation.signedBlock.encodedBlock
+            );
+            const coordinateKey = this.coordinatesToKey(block.coordinates);
+
+            this.hashToBlockMap.delete(hashOrForkId as Hash);
+            this.coordinatesToBlockMap.delete(coordinateKey);
+            return true;
+        }
+
+        // ┌─ ROUTES TO: [OVERLOAD 2] - delete by coordinates
+        const coordinateKey = this.coordinatesToKey({
+            forkId: hashOrForkId as ForkId,
+            height
+        });
+        const blockConfirmation = this.coordinatesToBlockMap.get(coordinateKey);
+        if (!blockConfirmation) return false;
+
+        // Need to find and delete from hash map too
+        const blockHash = ethers.keccak256(
+            blockConfirmation.signedBlock.encodedBlock
+        );
+
+        this.coordinatesToBlockMap.delete(coordinateKey);
+        this.hashToBlockMap.delete(blockHash);
+        return true;
+    }
+
+    // ====================================
+    // PRIVATE HELPERS
+    // ====================================
+
+    private coordinatesToKey(coordinates: BlockCoordinates): CoordinateKey {
+        return `${coordinates.forkId}:${coordinates.height}`;
+    }
+
+    private _storeBlockConfirmation(
+        blockConfirmation: BlockConfirmationStruct
+    ): Hash {
+        const blockHash = ethers.keccak256(
+            blockConfirmation.signedBlock.encodedBlock
+        );
+        const block = Block.decode(blockConfirmation.signedBlock.encodedBlock);
+
+        return this._storeBlockConfirmationWithKeys(
+            blockConfirmation,
+            blockHash,
+            block.coordinates
+        );
+    }
+
+    private _storeBlockConfirmationWithKeys(
+        blockConfirmation: BlockConfirmationStruct,
+        blockHash: Hash,
+        coordinates: BlockCoordinates
+    ): Hash {
+        const coordinateKey = this.coordinatesToKey(coordinates);
+
+        // Check for duplicates - throw if exists
+        if (this.hashToBlockMap.has(blockHash)) {
+            throw new Error(`Block with hash ${blockHash} already exists`);
+        }
+        if (this.coordinatesToBlockMap.has(coordinateKey)) {
+            throw new Error(
+                `Block at fork ${coordinates.forkId}, height ${coordinates.height} already exists`
```

Comment:
Probably also good to not error here - if the blockConfirmation exists already try and update it with signatures from the provided one and return the same hash regardless

Reply:
I don't understand why you want  these implicit behaviors

you don't want errors thrown? fine. that's a reasonable API design choice.
return key of prev existing entry? OK i guess, i think return something like `{success: boolean, key: Hash}` is better but maybe we don't need this added complexity.

but implicitly altering an existing entry?! why?

Reply:
No reason to error if it's already persisted - the system is not in an inconsistent state, it can progress - it can happen if a higher component has a race condition and is not prevented with something like a `mutex` - which in our case it is for now, but still 

As for my opinion on why to `try` and alter an exiting entry if an entry already exits, we're not breaking anything - we want to collect all the signatures and it shouldn't matter in which order we do it and if there's a race condition in a higher component and `storeBlock` is called multiple times - we can account for that easily with everything we already have here without exposing additional business logic